addons/stock/report/report_stock_forecasted.py

```diff
@@ -0,0 +1,207 @@
+            'product': {
+                'id': product.id,
+                'display_name': product.display_name
+            },
+            'replenishment_filled': replenishment_filled,
+            'uom_id': product.uom_id,
+            'receipt_date': format_datetime(self.env, move_in.date_expected, timezone, 'medium') if move_in else False,
+            'delivery_date': format_datetime(self.env, move_out.date_expected, timezone, 'medium') if move_out else False,
+            'receipt_date_short': format_date(self.env, move_in.date_expected) if move_in else False,
+            'delivery_date_short': format_date(self.env, move_out.date_expected) if move_out else False,
+            'is_late': is_late,
+            'quantity': quantity,
+            'move_out': move_out,
+            'move_in': move_in,
+        }
+
+    def _get_report_lines(self, product_template_ids, product_variant_ids, wh_location_ids):
+        in_domain, out_domain = self._move_confirmed_domain(
+            product_template_ids, product_variant_ids, wh_location_ids
+        )
+        outs = self.env['stock.move'].search(out_domain, order='date_expected, id')
+        outs_per_product = defaultdict(lambda: [])
+        for out in outs:
+            outs_per_product[out.product_id.id].append(out)
+        ins = self.env['stock.move'].search(in_domain, order='date_expected, id')
+        ins_per_product = defaultdict(lambda: [])
+        for in_ in ins:
+            ins_per_product[in_.product_id.id].append([in_.product_qty, in_])
+        currents = {c['id']: c['qty_available'] for c in outs.product_id.read(['qty_available'])}
+
+        lines = []
+        for product in (ins | outs).product_id:
+            for out in outs_per_product[product.id]:
+                # Reconcile with the current stock.
+                current = currents[out.product_id.id]
+                demand = out.product_qty
+                taken_from_stock = min(demand, current)
+                if not float_is_zero(taken_from_stock, precision_rounding=product.uom_id.rounding):
+                    currents[product.id] -= taken_from_stock
+                    demand -= taken_from_stock
+                    lines.append(self._prepare_report_line(taken_from_stock, move_out=out))
+                # Reconcile with the ins.
+                if not float_is_zero(demand, precision_rounding=product.uom_id.rounding):
+                    for index, in_ in enumerate(ins_per_product[out.product_id.id]):
+                        if float_is_zero(in_[0], precision_rounding=product.uom_id.rounding):
+                            continue
+                        taken_from_in = min(demand, in_[0])
+                        demand -= taken_from_in
+                        lines.append(self._prepare_report_line(taken_from_in, move_in=in_[1], move_out=out))
+                        ins_per_product[out.product_id.id][index][0] -= taken_from_in
```

Comment:
We could pop the item from the list. So it would avoid to browse it unnecessarily during next moves out